### ISSUE TITLE
Request for sixel support for image previews, to support multiplexers

### ISSUE THREAD

**Issue by AndydeCleyre** [Canop/broot#568]
While the terminal app I use supports the kitty image protocol, I am *always* in tmux, which decidedly will never support it.

~There is however a fork of tmux for sixel support~ **tmux now supports sixel**, and ~sixel support is imminent for zellij~ **zellij supports sixel as well**, which may be able to supplant tmux in my case eventually. Zellij, too, has no plans to ever support the kitty image protocol.

---

Another option to support multiplexers is ~the ueberzug x11 stuff~ **ueberzugpp** 🤷🏼.

**Comment by BluishHumility** [comment]
I wanted to +1 this request for Sixel support.

I have really been enjoying Broot. It is featureful and intuitive, and is out of the box better than I was able to get some other  TUI file explorers even after hours of patching and hacking around in config files. The only place it falls down for me is browsing image folders; the pixelated image previews are not enough for certain things. At the moment I typically escape from Broot into the normal terminal to run `lsix` (https://github.com/hackerb9/lsix) in order to view proper image thumbnails.

Adding Sixel support to Broot would massively expand the number of terminals that support previewing higher-resolution images. Here is a list (scooped off of the `lsix` site linked above):

SIXEL compatible terminals

- XTerm (tested)
- MLterm (tested)
- foot (tested)
- Wezterm (tested)
- Contour (tested)
- iTerm2 for Apple MacOS (tested)
- Konsole (reported)
- yakuake (reported)
- WSLtty for Microsoft Windows (reported)
- MinTTY for Cygwin (Microsoft Windows) (reported)
- Yaft for Linux framebuffer (tested)
- VTE (special compilation, reported)
- sixel-tmux (fork of tmux, reported)
- ttyd (reported)

I see this request is already nearly six months old, but I thought I would speak up and say there is still interest in having Sixel support in Broot. I really think it would be a valuable addition to this great project.

**Comment by Canop** [comment]
If somebody finds either a Rust code, something portable, or a precise specification for producing sixel escape codes from an image buffer, this would help this issue.
I don't want to import another native library like libsixel, and I'll only consider maintainable code.

**Comment by likewhatevs** [comment]
> If somebody finds either a Rust code, something portable, or a precise specification for producing sixel escape codes from an image buffer, this would help this issue.

I think [this function](https://github.com/o2sh/onefetch/blob/fdf435d089487a7b8eea7c6875bf6a282db2e4a8/image/src/sixel.rs#L82) is exactly that, it's in [this crate](https://crates.io/crates/onefetch-image). 

The key difference from the ideal (buffer in, escape codes out) is that it accepts `Vec<String>` and buffer instead of dimensions and buffer (i.e., it gets dimensions from `Vec<String>` and resizes accordingly).

Looking at [kitty support](https://github.com/Canop/broot/blob/main/src/kitty/image_renderer.rs), and that function in onefetch, I see all the components there (i.e. take things from image crate, resize to X lines, return a string of escape codes).

**Comment by Canop** [comment]
Thanks @likewhatevs, this is actually helpful!

**Comment by Canop** [comment]
> Adding Sixel support to Broot would massively expand the number of terminals that support previewing higher-resolution images

Not sure about that. On linux, is there even a usable terminal supporting sixel and not supporting the kitty protocol too (xterm can hardly count as it needs you to compile it yourself to add sixel support) ?

**Comment by BluishHumility** [comment]
> On linux, is there even a usable terminal supporting sixel and not supporting the kitty protocol too

I use Foot on Wayland machines, which was what motivated me to chime in on the thread. I guess how "usable" is defined here is relevant, but for what it's worth Foot is the default terminal for the Sway window manager.

I will admit I did not consider some of the Sixel-compatible terminals would also support the Kitty protocol, so it is likely I overstated the impact Sixel support will have.

**Comment by Efreak** [comment]
> On linux, is there even a usable terminal supporting sixel and not supporting the kitty protocol too (xterm can hardly count as it needs you to compile it yourself to add sixel support)

There's *many* terminals with sixel support, though I can't vouch for whether or not they support Kitty protocol. See https://www.arewesixelyet.com

**Comment by AndydeCleyre** [comment]
> On linux, is there even a usable terminal supporting sixel and not supporting the kitty protocol too

This is not a strong answer, because tmux doesn't support *either*, and the sixel-support tmux fork is not super maintained, but for me the answer is tmux. If broot supported sixel I would use the tmux fork...

Again, I see how this is not a strong answer.

**Comment by Canop** [comment]
Note that I have nothing against supporting sixel, but right now there's clear description of the protocol, and there's nothing reasonably clean and easy to port (and no pure Rust complete implementation).

**Comment by AndydeCleyre** [comment]
I think it's worth linking [ueberzugpp](https://github.com/jstkdng/ueberzugpp) here now:

```markdown
Überzug++ is a command line utility written in C++ which allows to draw images on terminals by using X11/wayland child windows, sixels, kitty and iterm2..

This project intends to be a drop-in replacement for the now defunct [ueberzug](https://github.com/seebye/ueberzug) project. If some tool doesn't work,
feel free to open an issue.

Advantages over w3mimgdisplay and ueberzug:

- support for wayland (sway and hyprland only) and MacOS
- no race conditions as a new window is created to display images
- expose events will be processed, so images will be redrawn on switch workspaces
- tmux support on X11, sway and hyprland
- terminals without the WINDOWID environment variable are supported
- chars are used as position - and size unit
- No memory leak (usage of smart pointers)
- A lot of image formats supported (through opencv and libvips).
- GIF and animated WEBP support on X11, Sixel, Sway and hyprland
- Fast image downscaling (through opencv and opencl)
- Cache resized images for faster viewing
```

**Comment by markus-bauer** [comment]
AFAIK, kitty now also supports placing images that should work in tmux:
https://sw.kovidgoyal.net/kitty/graphics-protocol/#unicode-placeholders
See also:
https://github.com/ranger/ranger/pull/2856

I haven't tested it though.

**Comment by wyndon** [comment]
+1

**Comment by n0bra1n3r** [comment]
How feasible would it be to support a "preview" verb? My idea is that this verb would be automatically called when a preview can be presented for the selected file, allowing the execution of an external command to preview the file:

```toml
[[verbs]]
  invocation = "preview"
  external = 'nvim --server $NVIM --remote-send "<C-\><C-N>:e {file}<CR>"'
  apply_to = "file"
```

I use broot in neovim as my file explorer, and this feature would allow me to show a file selected in broot as a vim buffer in the same instance. This would essentially allow broot to support all filetypes, highlighting and previews that vim supports.

I would set it up so that broot and the preview buffers appear in a single instance of neovim, but I imagine a similar setup could also be done with a combination of broot, tmux, and vim.

**Comment by BluishHumility** [comment]
> Note that I have nothing against supporting sixel, but right now there's no clear description of the protocol, and there's nothing reasonably clean and easy to port (and no pure Rust complete implementation).

I am not sure if it is reasonably clean or easy to port, but aside from a little bit of Lua the Yazi file manager is written in Rust, and they have incorporated Sixel support for image previewing in there somehow.

Understanding the code is significantly beyond my skill set, but grepping through the repo for "sixel" turns up this file: https://github.com/sxyazi/yazi/blob/main/yazi-adaptor/src/adaptor.rs and this one: https://github.com/sxyazi/yazi/blob/main/yazi-adaptor/src/lib.rs

I apologize if that is not helpful or relevant; like I mentioned I am not really sure what I am looking at but I thought I might mention it in case it is useful. I am still holding out hope for this feature! :crossed_fingers:

**Comment by BluishHumility** [comment]
> Yes! Konsole is one such terminal.

No, Konsole is not an example of such a terminal because it also supports the kitty protocol.

**Comment by AndydeCleyre** [comment]
I posted this on the wrong issue at first:

---

I just learned that official tmux can now be built with sixel support!

https://www.arewesixelyet.com/#tmux

**Comment by unisons** [comment]
It seems that on popup windows, the sixel is not working.

**Comment by stevenxxiu** [comment]
> If somebody finds either a Rust code, something portable, or a precise specification for producing sixel escape codes from an image buffer, this would help this issue. I don't want to import another native library like libsixel, and I'll only consider maintainable code.

I found the following *Rust* crates that don't depend on *libsixel*. The first one sounds better from the readme.

- [zellij-org/sixel-image: A lightning-fast Sixel serializer/deserializer](https://github.com/zellij-org/sixel-image)
- [mkrueger/icy_sixel: Port of libsixel to rust.](https://github.com/mkrueger/icy_sixel)

**Comment by AndydeCleyre** [comment]
Oh nice! I see an example at https://github.com/mkrueger/icy_sixel/pull/7

**Comment by Canop** [comment]
This looks like something to study, but a look at the quality (and maybe some contributions) will be needed. The project's README doesn't sound very definitive about it...

**Comment by ngocphamm** [comment]
> AFAIK, kitty now also supports placing images with unicode placeholders, which should work in tmux: https://sw.kovidgoyal.net/kitty/graphics-protocol/#unicode-placeholders See also: [ranger/ranger#2856](https://github.com/ranger/ranger/pull/2856)
> 
> I haven't tested it though.

I think this might be worth another look.

Yazi works inside tmux now, I think because of that.

My setup
* macOS 15.2
* [Ghostty](https://github.com/ghostty-org/ghostty) 1.0
* Yazi 0.4.2
* tmux 3.5a (from `brew install tmux`)
    * I don't even have to apply the 3 lines as mentioned in the Yazi [guide](https://yazi-rs.github.io/docs/image-preview#tmux). I'm not sure if that's because of my tmux version or what.

**Comment by AndydeCleyre** [comment]
Yes, I think if broot supported this unicode placeholder method, now that ghostty which supports it is public (and wezterm will probably eventually support it), I would be happy to withdraw my request for sixel support.

**Comment by AndydeCleyre** [comment]
Is anyone subscribed to this issue still set on sixel support for some reason, given that now there's a way to implement the kitty image protocol such that it works with tmux? 

If not, I'm inclined to morph this issue into a request for kitty's unicode placeholder system, as my real need is image support in tmux, and this would accomplish that.

@Canop In addition to the [kitty docs on the matter](https://sw.kovidgoyal.net/kitty/graphics-protocol/#unicode-placeholders), here's [yazi's implementation in rust](https://github.com/sxyazi/yazi/blob/main/yazi-adapter/src/drivers/kgp.rs).

**Comment by Canop** [comment]
I'd clearly prefer to have distinct issues. This one is about sixel support and anything not related to sixel here is treated as mere informative comments, not as real requests, until it's moved elsewhere.

**Comment by j4james** [comment]
> > Yes! Konsole is one such terminal.
> 
> No, Konsole is not an example of such a terminal because it also supports the kitty protocol.

Note that Konsole doesn't support [Unicode placeholders](https://sw.kovidgoyal.net/kitty/graphics-protocol/#unicode-placeholders), without which you're not going to get images working in tmux, which is the main point of this issue. So if you're only supporting the kitty graphics protocol, then there are only two terminals I'm aware of that'll work in tmux - kitty and ghostty.

However Sixel is supported by loads of terminals (I've personally tested around 20, but there's definitely more than that). The quality of that support will vary, so it's possible they don't all work in tmux, but I'm willing to bet there are lot more than two that will work.

**Comment by lucaspimentel** [comment]
It's not Linux, but [Windows Terminal](https://github.com/microsoft/terminal) is another terminal that supports the Sixel image protocol but not the Kitty protocol. Windows Terminal is quickly the default terminal on Windows (at least for advanced users, developers, etc), and is the terminal that [the broot documentation](https://dystroy.org/broot/install/) recommends for Windows users:

![Image](https://github.com/user-attachments/assets/d2823e24-896f-43e4-85b1-6dc13f2ee407)

**Comment by jamison-wilde** [comment]
[PR for Sixel Support](https://github.com/Canop/broot/pull/1175/)
Tested in Windows Terminal and Ubuntu xterm.

**Comment by Canop** [comment]
People interested in sixel's integration: I'd need your feedback on @jamison-wilde 's PR

**Comment by xbwwj** [comment]
Works perfectly in Foot terminal, iTerm2, and Termux's [merging sixel branch](https://github.com/termux/termux-app/pull/2973). Thank you so much for implementing this! @jamison-wilde

**Comment by Canop** [comment]
Sixel support shipped in broot 1.59 (thanks @jamison-wilde, and everybody who tested).

The official binaries are built with it, and outside a multiplexer it's selected automatically when the kitty protocol isn't available. If you compile broot yourself, enable the feature (it needs rust 1.90+):

    cargo install --locked --features clipboard,sixel broot

In tmux, auto-detection can't see the outer terminal, so set `graphics_display: sixel` in the configuration, or run with `BROOT_GRAPHICS_PROTOCOL=sixel`. See https://dystroy.org/broot/common-problems/#hi-res-images-in-tmux-sixel

Please open a new issue for any terminal where detection fails.
